File: coding/src/com/algos/core/li01_array/revised/PrefixSum_CountNumberOfSubarraysWithXorK.java
package com.algos.core.li01_array.revised;

import java.util.HashMap;
import java.util.Map;

public class PrefixSum_CountNumberOfSubarraysWithXorK {
    
        public static void main(String[] args) {
            PrefixSum_CountNumberOfSubarraysWithXorK obj = new PrefixSum_CountNumberOfSubarraysWithXorK();
            
            int[] nums = new int[] {1,1,2,3,4};
            System.out.println(obj.subarrayXorK(nums, 3));
        }
    
        // Idea: Use a map to store prefix sum and its frequency
        // Same as count sum arrays with sum k
        // XR: xor(0 -> curr) and y: xor(0 -> j) where j < curr
        // x ^ k = XR       
        // x = XR ^ k
        // Time: O(n), Space: O(n)
        public int subarrayXorK(int[] nums, int k) {
            int XR = 0, cnt = 0;
            Map<Integer, Integer> map = new HashMap<>();
            map.put(0, 1);
            for(int num: nums){
                XR ^= num;
                int x = XR^k;
                if (map.containsKey(x)) {
                    cnt += map.get(x);
                }
                map.putIfAbsent(XR, 0);
                map.put(XR, map.get(XR)+1);
            }
            
            return cnt;
        }     
}

File: dsa/concepts/01_arrays/python/prefix_sum_count_number_of_subarrays_with_xor_k.py
"""
PrefixSum - Count Number of Subarrays with XOR K

Idea: Use a map to store prefix sum and its frequency
Same as count sum arrays with sum k
XR: xor(0 -> curr) and y: xor(0 -> j) where j < curr
x ^ k = XR
x = XR ^ k
Time: O(n), Space: O(n)
"""

def subarray_xor_k(nums, k):
    xr = 0
    cnt = 0
    map_freq = {0: 1}
    
    for num in nums:
        xr ^= num
        x = xr ^ k
        if x in map_freq:
            cnt += map_freq[x]
        
        if xr not in map_freq:
            map_freq[xr] = 0
        map_freq[xr] += 1
    
    return cnt


if __name__ == "__main__":
    nums = [1, 1, 2, 3, 4]
    print(subarray_xor_k(nums, 3))
